build.sbt

```diff
@@ -103,53 +105,77 @@ lazy val commonSettings = Seq(
   // format: on
 )
 
-// add support for Scala version ranges such as "scala-2.12+" in source folders (single version folders such as
-// "scala-2.11" are natively supported by SBT).
-// In order to keep this simple, we're doing this case by case, taking advantage of the fact that we intend to support
-// only 3 major versions at any given moment.
+// add support for Scala version ranges such as "scala-2.12+" or "scala-2.13-" in source folders (single version folders
+// such as "scala-2.11" are natively supported by SBT).
 def crossVersionSharedSources(unmanagedSrcs: SettingKey[Seq[File]]) = {
   unmanagedSrcs ++= {
-    val minor = CrossVersion.partialVersion(scalaVersion.value).map(_._2)
-    List(
-      if (minor.exists(_ <= 12)) unmanagedSrcs.value.map { dir => new File(dir.getPath + "-2.12-") }
-      else Nil,
-      if (minor.exists(_ >= 12)) unmanagedSrcs.value.map { dir => new File(dir.getPath + "-2.12+") }
-      else Nil
-    ).flatten
+    val versionNumber = CrossVersion.partialVersion(scalaVersion.value)
+    val expectedVersions = Seq(scala211, scala212, scala213, scala30).flatMap(CrossVersion.partialVersion)
+    expectedVersions.flatMap { case v @ (major, minor) =>
+      List(
+        if (versionNumber.exists(Ordering[(Long, Long)].lteq(_, v)))
+          unmanagedSrcs.value.map { dir => new File(dir.getPath + s"-$major.$minor-") }
+        else
+          Nil,
+        if (versionNumber.exists(Ordering[(Long, Long)].gteq(_, v)))
+          unmanagedSrcs.value.map { dir => new File(dir.getPath + s"-$major.$minor+") }
+        else
+          Nil
+      )
+    }.flatten
   }
 }
 
-lazy val lintFlags = {
-  lazy val allVersionLintFlags = List(
-    "-encoding",
-    "UTF-8", // arg for -encoding
-    "-feature",
-    "-unchecked",
-    "-Ywarn-dead-code",
-    "-Ywarn-numeric-widen"
-  )
-
-  def withCommon(flags: String*) =
-    allVersionLintFlags ++ flags
-
-  forScalaVersions {
+lazy val lintFlags = forScalaVersions { v =>
+  (v: @unchecked) match {
```

Comment:
How about getting rid of this `v` and using a partial function with `forScalaVersions` directly?

Reply:
I think we lose the selector to put the `@unchecked` annotation that way. The warning is a bit annoying.

Reply:
I may be missing some detail, but the annotation doesn't seem to be needed here as you're not pattern-matching on any erased type. If you start pattern matching directly you may need to add a new `case _ => Nil` or `case _ => throw new Exception("Unknown version")`, which is good practice anyway, but other than that everything should work fine without warnings.

Reply:
Right, sorry! I was assuming you meant using a partial function directly. Adding a fallback works as well, yes.